File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/screen/base/adapters/BaseAdapter.java
package ru.avystavkin.githubgists.screen.base.adapters;

import android.view.View;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import ru.avystavkin.githubgists.screen.interfaces.OnItemClickListener;

public abstract class BaseAdapter<VH extends RecyclerView.ViewHolder, T> extends BaseRecyclerViewAdapter<VH, T> {

    private final List<T> mItems = new ArrayList<>();

    @Nullable
    private OnItemClickListener<T> mOnItemClickListener;

    private final View.OnClickListener mInternalListener = (view) -> {
        if (mOnItemClickListener != null) {
            int position = (int) view.getTag();
            T item = mItems.get(position);
            mOnItemClickListener.onItemClick(item);
        }
    };

    public BaseAdapter(@NonNull List<T> items) {
        mItems.addAll(items);
    }

    public final void add(@NonNull T value) {
        mItems.add(value);
        refreshRecycler();
    }

    public final void changeDataSet(@NonNull List<T> values) {
        mItems.clear();
        mItems.addAll(values);
        refreshRecycler();
    }

    public final void clear() {
        mItems.clear();
        refreshRecycler();
    }


    @CallSuper
    @Override
    public void onBindViewHolder(VH holder, int position) {
        holder.itemView.setTag(position);
        holder.itemView.setOnClickListener(mInternalListener);
    }

    public void setOnItemClickListener(@Nullable OnItemClickListener<T> onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    @NonNull
    public T getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }
}

File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/screen/base/adapters/BaseAdapter.kt
package ru.avystavkin.githubgists.screen.base.adapters

import android.view.View
import androidx.annotation.CallSuper
import androidx.recyclerview.widget.RecyclerView
import ru.avystavkin.githubgists.screen.interfaces.OnItemClickListener
import java.util.*

abstract class BaseAdapter<VH : RecyclerView.ViewHolder, T>(items: List<T>) : BaseRecyclerViewAdapter<VH>() {

    private val mItems = ArrayList<T>(items)

    var onItemClickListener: OnItemClickListener<T>? = null

    private val mInternalListener = View.OnClickListener { view ->
        if (onItemClickListener != null) {
            val position = view.tag as Int
            val item = mItems[position]
            onItemClickListener!!.onItemClick(item)
        }
    }

    fun add(value: T) {
        mItems.add(value)
        refreshRecycler()
    }

    fun changeDataSet(values: List<T>) {
        mItems.clear()
        mItems.addAll(values)
        refreshRecycler()
    }

    fun clear() {
        mItems.clear()
        refreshRecycler()
    }


    @CallSuper
    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.itemView.tag = position
        holder.itemView.setOnClickListener(mInternalListener)
    }

    fun getItem(position: Int): T {
        return mItems[position]
    }

    override fun getItemCount(): Int {
        return mItems.size
    }
}
